Introduce `wp-on-async` directive as performant alternative over synchronous `wp-on` directive

westonruter opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

Fixes #61634.

## What?

This adds an alternative async variation to the `data-wp-on` directives by tagging on an `-async` to the directive name. So instead of `data-wp-on--click` it can be `data-wp-on-async--click`. This is also done for the `data-wp-on-window` and `data-wp-on-document` directives with `data-wp-on-async-window` and`data-wp-on-async-document` respectively (which here can also be made `passive`).

## Why?

This ensures that when there are multiple directives for the same event, they do not add up to a long task.

## How?

This directive yields to the main thread immediately before invoking the action/callback.

This is implemented for the eligible directives in the Image block. 

<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions

1. Install and activate this [Try Slow Events](https://gist.github.com/westonruter/59f7c58230bd5d1e82733e5b87d245dc) plugin which adds 10 click event directives to the `img` in an Image block.
2. Add an Image to a post and enable the click-to-expand feature.
3. Navigate to the post on the frontend.
4. First add `?try-slow-events=sync` to the URL and try opening the image in a lightbox. Notice the latency.
5. Second, try `?try-slow-events=async` to the URL and try opening the image in a lightbox. Notice the latency goes away.

## Screenshots or screencast <!-- if applicable -->

### Before (`?try-slow-events=sync`)

[Screen recording 2024-05-22 17.34.29.webm](https://github.com/WordPress/gutenberg/assets/134745/5ffb9ab4-1706-44f7-989f-812b8c8a35d2)

![image](https://github.com/WordPress/gutenberg/assets/134745/a14c546f-210b-4a9a-a884-91eab25bd8d2)

### After (`?try-slow-events=async`)

[Screen recording 2024-05-22 17.34.52.webm](https://github.com/WordPress/gutenberg/assets/134745/19457cf1-a9d0-43a3-8f8e-731007be6cf4)

![image](https://github.com/WordPress/gutenberg/assets/134745/a35633e4-ae03-44e9-979d-e452696da84a)

(Aside: The `yieldToMain()` seems it could be improved yet further since a long task is still sneaking in, even though it is much better.)

westonruter committed:
Improve typing for data-wp-on directive

westonruter committed:
Move yieldToMain to utils

westonruter committed:
fixup! Improve typing for data-wp-on directive

westonruter committed:
Introduce on-async directives

westonruter committed:
Use on-async directives in image block

cbravobernal commented:
Thanks @westonruter for the PR!

My main concern is that developers won't use the async mode as such as needed.

What could be the implications of making it async by default and switching to `wp-on-sync-` for the rest of cases?

How can we guide developers for the best approach, apart from updating the docs?

cbravobernal reviewed (dismissed):
Code looks good, and tests are working as expected.

It would be nice to add some e2e or check if we can update the ones we have for the document and window directives.

We still need to update the documentation to include these new directives.

cbravobernal committed:
Update changelog

westonruter commented:
> My main concern is that developers won't use the async mode as such as needed.
> 
> What could be the implications of making it async by default and switching to `wp-on-sync-` for the rest of cases?

That would be ideal, if the back-compat breakage is acceptable. Basically, synchronous events are required when passing in an `event` object _and_ when the handler calls `event.preventDefault()`, `event.stopPropagation()`, or `event.stopImmediatePropagation()`. Two of the actions on the Image block call `event.preventDefault()` (as listed in #61634). For example, if `actions.handleKeydown` were to be made async, then no longer would Tab key presses get trapped in the opened lightbox since `event.preventDefault()` would be called too late.

> How can we guide developers for the best approach, apart from updating the docs?

One possibility is that we warn users about switching to use `on-async` when the associated action lacks any argument. Perhaps when `SCRIPT_DEBUG` is enabled, when looping over the directives to add to the `events` Map, it could look up the referenced function and stringify it to then parse out the function's argument list. If it is empty and `on` is used, it could warn that the user should use `on-async` instead.

westonruter committed:
Fix typo in comment

westonruter commented:
> One possibility is that we warn users about switching to use `on-async` when the associated action lacks any argument. Perhaps when `SCRIPT_DEBUG` is enabled, when looping over the directives to add to the `events` Map, it could look up the referenced function and stringify it to then parse out the function's argument list. If it is empty and `on` is used, it could warn that the user should use `on-async` instead. (Maybe the function body could also be looked at to see if it contains `preventDefault`, `stopPropagation`, or `stopImmediatePropagation`.)

Humm. This doesn't seem to work as I expected because the actions/callbacks are getting wrapped in additional functions.

<details><summary>My failed patch</summary>

```diff
diff --git a/packages/interactivity/src/directives.tsx b/packages/interactivity/src/directives.tsx
index a4a320eda57..7af341da558 100644
--- a/packages/interactivity/src/directives.tsx
+++ b/packages/interactivity/src/directives.tsx
@@ -19,7 +19,7 @@ import {
 	yieldToMain,
 } from './utils';
 import type { DirectiveEntry } from './hooks';
-import { directive, getScope, getEvaluate } from './hooks';
+import { directive, getScope, getEvaluate, getDereference } from './hooks';
 
 // Assigned objects should be ignored during proxification.
 const contextAssignedObjects = new WeakMap();
@@ -314,7 +314,7 @@ export default () => {
 	} );
 
 	// data-wp-on--[event]
-	directive( 'on', ( { directives: { on }, element, evaluate } ) => {
+	directive( 'on', ( { directives: { on }, element, evaluate, dereference } ) => {
 		const events = new Map< string, Set< DirectiveEntry > >();
 		on.filter( ( { suffix } ) => suffix !== 'default' ).forEach(
 			( entry ) => {
@@ -328,6 +328,14 @@ export default () => {
 
 		events.forEach( ( entries, eventType ) => {
 			const existingHandler = element.props[ `on${ eventType }` ];
+
+			if ( typeof SCRIPT_DEBUG === 'boolean' && SCRIPT_DEBUG ) {
+				for ( const entry of entries ) {
+					const handler = dereference( entry ).value;
+					console.info(handler.toString());
+				}
+			}
+
 			element.props[ `on${ eventType }` ] = ( event: Event ) => {
 				entries.forEach( ( entry ) => {
 					if ( existingHandler ) {
diff --git a/packages/interactivity/src/hooks.tsx b/packages/interactivity/src/hooks.tsx
index d8aa43d2c0e..9fc403c7801 100644
--- a/packages/interactivity/src/hooks.tsx
+++ b/packages/interactivity/src/hooks.tsx
@@ -53,6 +53,11 @@ interface DirectiveArgs {
 	 * context.
 	 */
 	evaluate: Evaluate;
+	/**
+	 * Function that dereferences a given path to a value either in the store or the
+	 * context.
+	 */
+	dereference: Dereference;
 }
 
 interface DirectiveCallback {
@@ -71,6 +76,7 @@ interface DirectiveOptions {
 
 interface Scope {
 	evaluate: Evaluate;
+	dereference: Dereference;
 	context: object;
 	ref: RefObject< HTMLElement >;
 	attributes: createElement.JSX.HTMLAttributes;
@@ -80,10 +86,18 @@ interface Evaluate {
 	( entry: DirectiveEntry, ...args: any[] ): any;
 }
 
+interface Dereference {
+	( entry: DirectiveEntry ): any;
+}
+
 interface GetEvaluate {
 	( args: { scope: Scope } ): Evaluate;
 }
 
+interface GetDereference {
+	( args: { scope: Scope } ): Dereference;
+}
+
 type PriorityLevel = string[];
 
 interface GetPriorityLevels {
@@ -307,6 +321,24 @@ export const getEvaluate: GetEvaluate =
 		return hasNegationOperator ? ! result : result;
 	};
 
+// Generate the dereference function.
+export const getDereference: GetDereference =
+	( { scope } ) =>
+	( entry ) => {
+		let { value: path, namespace } = entry;
+		if ( typeof path !== 'string' ) {
+			throw new Error( 'The `value` prop should be a string path' );
+		}
+		// If path starts with !, remove it and save a flag.
+		const hasNegationOperator =
+			path[ 0 ] === '!' && !! ( path = path.slice( 1 ) );
+		setScope( scope );
+		const value = resolve( path, namespace );
+		const result = { value, hasNegationOperator };
+		resetScope();
+		return result;
+	};
+
 // Separate directives by priority. The resulting array contains objects
 // of directives grouped by same priority, and sorted in ascending order.
 const getPriorityLevels: GetPriorityLevels = ( directives ) => {
@@ -338,6 +370,7 @@ const Directives = ( {
 	// element ref, state and props.
 	const scope = useRef< Scope >( {} as Scope ).current;
 	scope.evaluate = useCallback( getEvaluate( { scope } ), [] );
+	scope.dereference = useCallback( getDereference( { scope } ), [] );
 	scope.context = useContext( context );
 	/* eslint-disable react-hooks/rules-of-hooks */
 	scope.ref = previousScope?.ref || useRef( null );
@@ -367,6 +400,7 @@ const Directives = ( {
 		element,
 		context,
 		evaluate: scope.evaluate,
+		dereference: scope.dereference,
 	};
 
 	setScope( scope );
```

</details> 

When dereferencing a function, I'm seeing its source as a closure coming from `handlers` in `store.ts`:

```js
(...args) => {
  (0,_hooks__WEBPACK_IMPORTED_MODULE_1__.setNamespace)(ns);
  try {
    return result(...args);
  } finally {
    (0,_hooks__WEBPACK_IMPORTED_MODULE_1__.resetNamespace)();
  }
}
```

cbravobernal commented:
We only have this week to land this change before beta 1.

Should we merge this, go with the `data-wp-on-async` directives and work later on the warning @westonruter ?

westonruter commented:
@cbravobernal That makes sense to me!

gziolo commented:
Let's remember to document the new API. It should be very similar to:

https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/interactivity-api/api-reference.md#wp-on

cbravobernal commented:
I fixed the changelog conflict.

adamsilverstein commented:
> Let's remember to document the new API. It should be very similar to:
> 
> [`trunk`/docs/reference-guides/interactivity-api/api-reference.md#wp-on](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/interactivity-api/api-reference.md?rgh-link-date=2024-05-28T10%3A36%3A06Z#wp-on)
> 
> It would be also worth explaining the nuances and when to use `wp-on-async` instead of `wp-on`. The way I read it is - always use the async version when there is no `event` consumed.

I added the "Needs Dev Note label here - this feature is worth calling out for the 6.6 core release.

westonruter commented:
I've opened #62160 to implement `data-wp-on-async` for the other core blocks.

cbravobernal commented:
> Dev note

WordPress 6.6 includes three new directives aimed at performance optimization:
 - `data-wp-on-async`
 - `data-wp-on-document-async`
 - `data-wp-on-window-async`
 
These `async` optimize event callbacks by yielding them to the main thread. Thit way, complex interactions won't block the main thread, improving the INP. ([Interaction to next paint](https://web.dev/articles/inp)). You can read more about it at https://web.dev/articles/optimize-inp#yield-to-allow-rendering-work-to-occur-sooner

These directives are recommended over the sync ones, but you can use them only when there is no `event` consumed.
If your code is using the event object, such as in `event.stopPropagation()` or `event.preventDefault(), use the synchronous versions (like `data-wp-on`) instead.

westonruter commented:
I just realized that I didn't export `yieldToMain`. Proposed in https://github.com/WordPress/gutenberg/pull/62665

cbravobernal commented:
> @cbravobernal Thanks for drafting that dev note! I've made some edits.

I initially put that way and GTP4 corrected me to this way 😆

westonruter commented:
@cbravobernal I've updated the dev note comment above to incorporate discussion of the new `splitTask` function and to fix the names of the directives (also updated in #62759).

janusqa commented:
I've been looking at the docs here:
https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/api-reference/

Hi there.  I am using wp-on-window--message to listen for messages sent from an IFRAME to the parent window.
wp-on-window--message seems to work ok.  I get the message, but when i switch over to wp-on-async-window--message no messages arrive.

I've also read this from the PR. Does it mean its too early for me to be using this? The docs indicate 6.5 or higher.
```
WordPress 6.6 includes three new directives aimed at improving performance:

    data-wp-on-async
    data-wp-on-async-document
    data-wp-on-async-window
```

The particulars:
```
wordpress 6.5.5
"@wordpress/interactivity": "^6.3.0",
"@wordpress/scripts": "^27.9.0",
```

Hi there.  I am using wp-on-window--message to listen for messages sent from an IFRAME to the parent window.
wp-on-window--message seems to work ok.  I get the message, but when i switch over to wp-on-async-window--message no messages arrive.

The basics

render.php
```
<div data-wp-on-async-window--message="actions.processPostMessage" id="<?php echo $block_unique_id ?>" class="the-view" data-wp-interactive="create-block" <?php echo wp_interactivity_data_wp_context($block_context) ?>>
<!-- some html goes here -->
</div>
<iframe class="iframe-container" name="theFrame"></iframe>
```

view.js
```
/**
 * WordPress dependencies
 */
import { store, getContext } from '@wordpress/interactivity';
import { longRunningAsyncFunction } from './thefunction';

store('create-block', {
    actions: {
        processPostMessage: function* (event) {
            if (event.origin !== 'https://test.ipg-online.com') return;
            const context = getContext();
            event.data.elementArr.push({ theThing: context.the_thing });
            yield longRunningAsyncFunction(event);
        },
    },
    callbacks: {},
});
```
